### DISCUSSION TITLE
Wall Material Layers Volume

### DISCUSSION THREAD

**Discussion by AfxentiouNikolas** [IfcOpenShell/IfcOpenShell#3188] [Q&A]
Hello, I am new to IFC Open Shell, and I am trying to find the volume for each material layer. I am working with the schema: IFC4 Design Transfer View. I can extract the material layers and thickness for all walls. Here is my current result output:

![results](https://github.com/IfcOpenShell/IfcOpenShell/assets/97607951/1ae499cb-8fb8-4163-9c10-d49b48b5501e)

I would like to know how I can also extract the length and height so that I can calculate the volume for each material. Alternatively, is there a way to directly extract the volume for each material?

Best regards,
Nikolas

**Comment by aothms** [comment]
Not everything can be described as thickness x length x height. Walls can have openings, or be clipped to the roof, ...

One way of doing this is to use `ifcopenshell.geom.create_shape(ifcopenshell.geom.settings(APPLY_LAYERSETS=True), ...)`. That way you have IfcOpenShell construct an actual geometric surface from the layerset boundaries to subdivide the geometry. And then you can simply calculate quantities off of that.

Otherwise, length you can get by taking the 'Axis' representation and calculating the length (should be available quite generally on walls).

Height maybe just by taking top vertex and subtract bottom vertex z.

In all these cases I would recommend Python OpenCASCADE. Use TopoDS_Iterator to explore the subshapes for individual layers in the first case. BRepGProp_LinearProperties for edge lengths. TopExp_Explorer(TopAbs_VERTEX) BRep_Tool_Pnt() for exploring vertices and obtaining its point.
